File: dsa-problems/Stacks/solutions/javascript/STK-002-lab-mixed-bracket-repair.js
class Solution {
  countChanges(s) {
    const stack = [];
    const opens = "([{";
    const closes = ")]}";
    const pairs = { ')': '(', ']': '[', '}': '{' };

    for (const c of s) {
      if (opens.includes(c)) {
        stack.push(c);
      } else if (closes.includes(c)) {
        if (stack.length > 0 && stack[stack.length - 1] === pairs[c]) {
          stack.pop();
        } else {
          stack.push(c);
        }
      } else if (c === '?') {
        stack.push('(');
      }
    }
    return stack.length;
  }
}

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = "";
rl.on("line", (line) => {
  data += line;
});

rl.on("close", () => {
  const s = data.trim();
  if (s.length === 0) {
      console.log(0);
      return;
  }
  const solution = new Solution();
  console.log(solution.countChanges(s));
});

File: dsa-problems/Stacks/solutions/java/STK-002-lab-mixed-bracket-repair.java
import java.util.*;
import java.io.*;

class Solution {
    public int countChanges(String s) {
        Stack<Character> stack = new Stack<>();
        String opens = "([{";
        String closes = ")]}";
        Map<Character, Character> pairs = new HashMap<>();
        pairs.put(')', '(');
        pairs.put(']', '[');
        pairs.put('}', '{');
        
        for (char c : s.toCharArray()) {
            if (opens.indexOf(c) != -1) {
                stack.push(c);
            } else if (closes.indexOf(c) != -1) {
                if (!stack.isEmpty() && stack.peek() == pairs.get(c)) {
                    stack.pop();
                } else {
                    stack.push(c);
                }
            } else if (c == '?') {
                stack.push('(');
            }
        }
        
        return stack.size();
    }
}

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = br.readLine()) != null) {
            sb.append(line);
        }
        String s = sb.toString().trim();
        
        if (s.isEmpty()) {
            System.out.println(0);
            return;
        }

        Solution sol = new Solution();
        System.out.println(sol.countChanges(s));
    }
}
